src/SAWScript/Position.hs

```diff
@@ -30,12 +30,49 @@ import qualified What4.FunctionName as W4
 
 -- Pos ------------------------------------------------------------------------
 
+-- Type inference info, to be used to interpret the positions of
+-- inferred types.
+--
+-- InfFresh means that the term (or pattern) at the given position
+--    caused us to generate a fresh type variable, e.g. the element
+--    type of "[]".
+-- InfTerm means that the term (or pattern) at the given position
+--    prompted us to choose the accompanying type; e.g. "[]" has type
+--    List.
+-- InfContext means that the usage of the term at the given position
+--    prompted us to choose the accompanying type; e.g. in "f x" f has
+--    function type.
+data Inference
+  = InfFresh
+  | InfTerm
+  | InfContext
+  deriving (Data, Eq)
+
+-- Source position.
+--
+-- Mostly, this is a physical range within a file.
+--
+-- XXX: some logic uses Range with zero line/column numbers for end-of-file.
+-- This should be replaced with an explicit end-of-file position.
+--
+-- Internally generated objects use PosInternal with descriptive text.
+-- (FUTURE: eliminate PosInternal in favor of explicit constructors
+-- for the cases that currently generate PosInternal, so we can keep
+-- track of what they are.)
+--
+-- PosREPL is for things typed at the repl, or in some cases (that
+-- should get cleaned out) for things we assume came from the repl.
+--
+-- PosInferred is used for types generated by type inference; it
+-- describes where and how the inference was made. This can be thought
+-- of as the provenance of the inferred type.
```

Comment:
Similarly, note that `Pos` deliberately does _not_ have an `Ord` instance, as it is meant to be compared using `comparePosQuality` instead.

Reply:
Same